lib/nexthop.c

```diff
@@ -36,40 +36,94 @@
 DEFINE_MTYPE_STATIC(LIB, NEXTHOP, "Nexthop")
 DEFINE_MTYPE_STATIC(LIB, NH_LABEL, "Nexthop label")
 
-/* check if nexthops are same, non-recursive */
-int nexthop_same_no_recurse(const struct nexthop *next1,
-			    const struct nexthop *next2)
+static int nexthop_labels_cmp(const struct nexthop *nh1,
+			      const struct nexthop *nh2)
 {
-	if (next1->type != next2->type)
+	const struct mpls_label_stack *nhl1 = NULL;
+	const struct mpls_label_stack *nhl2 = NULL;
+
+	nhl1 = nh1->nh_label;
+	nhl2 = nh2->nh_label;
+
+	/* No labels is a match */
+	if (!nhl1 && !nhl2)
 		return 0;
 
+	if (nhl1 && !nhl2)
+		return 1;
+
+	if (nhl2 && !nhl1)
+		return -1;
+
+	if (nhl1->num_labels > nhl2->num_labels)
+		return 1;
+
+	if (nhl1->num_labels < nhl2->num_labels)
+		return -1;
+
+	return memcmp(nhl1->label, nhl2->label, nhl1->num_labels);
+}
+
+int nexthop_cmp(const struct nexthop *next1, const struct nexthop *next2)
+{
+	int ret = 0;
+	uint32_t n1 = 0;
+	uint32_t n2 = 0;
+
+	if (next1->vrf_id < next2->vrf_id)
+		return -1;
+
+	if (next1->vrf_id > next2->vrf_id)
+		return 1;
+
+	if (next1->type < next2->type)
+		return -1;
+
+	if (next1->type > next2->type)
+		return 1;
+
 	switch (next1->type) {
 	case NEXTHOP_TYPE_IPV4:
-	case NEXTHOP_TYPE_IPV4_IFINDEX:
-		if (!IPV4_ADDR_SAME(&next1->gate.ipv4, &next2->gate.ipv4))
-			return 0;
-		if (next1->ifindex && (next1->ifindex != next2->ifindex))
-			return 0;
-		break;
-	case NEXTHOP_TYPE_IFINDEX:
-		if (next1->ifindex != next2->ifindex)
-			return 0;
+		n1 = ntohl(next1->gate.ipv4.s_addr);
+		n2 = ntohl(next2->gate.ipv4.s_addr);
+		if (n1 < n2)
+			return -1;
+		if (n1 > n2)
+			return 1;
 		break;
 	case NEXTHOP_TYPE_IPV6:
-		if (!IPV6_ADDR_SAME(&next1->gate.ipv6, &next2->gate.ipv6))
-			return 0;
+		ret = memcmp(&next1->gate, &next2->gate, sizeof(union g_addr));
+		if (ret)
+			return ret;
 		break;
+	case NEXTHOP_TYPE_IPV4_IFINDEX:
 	case NEXTHOP_TYPE_IPV6_IFINDEX:
-		if (!IPV6_ADDR_SAME(&next1->gate.ipv6, &next2->gate.ipv6))
-			return 0;
-		if (next1->ifindex != next2->ifindex)
-			return 0;
+		ret = memcmp(&next1->gate, &next2->gate, sizeof(union g_addr));
```

Comment:
I don't like this very much since we're doing a full 16-byte comparison even for IPv4 addresses, and for that we assume that the g_addr union was zero-initialized. But this should be fine since we always create nexthops using `nexthop_new()`, which allocates memory using `XCALLOC`.